### PR TITLE
layer: Add max_events_per_span API

### PR DESCRIPTION
**Pull request by do-not-use-parker-timmerman**
## Motivation
As described in https://github.com/MaterializeInc/database-issues/issues/5082, there is a slow OOM when enabling open-telemetry tracing at the debug level. Through some investigation it turns out this OOM is caused by an underlying library, `h2`, having a long running span. Specifically, `h2` keeps a debug span open for the entire lifetime of a `Connection`. This is problematic because the `OpenTelemetryLayer` continuously appends events onto this span, and we don't flush them until the span closes, which is never does, so we end up using more and more memory until we OOM.

## Solution
This PR adds a new configurable field to `OpenTelemetryLayer` that allows you to configure the maximum number of `otel::Event`s that will be recorded for a single span. When we go over this max we drop the oldest events, and report that events were dropped.

### CODE REVIEW

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs]: a tracing event inside a tracing layer is a recipe for disaster i think...

In the context of mz, an `eprintln!` might work, but getting that merged upstream is impossible...can we instead fabricate one extra otel event and attach it to the otel span `on_close`?

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs:622]: this being configurable is great!

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs:952]: curious, whats the motivation here over say, just saving the most recent events?

**Review by guswynn — COMMENTED** [review]
Looks pretty good, I just have a few comments! After those comments, I'd definitely attempt to pr this against upstream!

File: tracing-opentelemetry/src/layer.rs
Comment (do-not-use-parker-timmerman) [reply tracing-opentelemetry/src/layer.rs]: Yeah I was definitely weary of this 😅

I changed the impl to add an attribute during `on_close`, that's what we seemed to do with `Timings`, what do you think about that?

File: tracing-opentelemetry/src/layer.rs
Comment (do-not-use-parker-timmerman) [reply tracing-opentelemetry/src/layer.rs:952]: What do you mean by "saving" the most recent events?

I'm pretty sure what this should be doing is dropping the oldest events first, but not all of the events. If we dropped all of the events every time what I was concerned about is the corner case where we'd get just one more than your max, and so we'd always end up dropping a ton of seemingly useful data.

A different strategy we could take is always popping the first event off the vec, and appending one to the back. This would be ideal but given the way `Vec` is implemented it wouldn't be that performant. Ideally `events` is a `VecDeque` and then we could do this

I tried to get the best of both worlds by this "drop half" strategy. We make way for more events, but not too many, and by using `.drain(...)` we re-use the underlying buffer from the `Vec`

**Comment by do-not-use-parker-timmerman** [comment]
> Looks pretty good, I just have a few comments! After those comments, I'd definitely attempt to pr this against upstream!

Thanks! Def will try to upstream this after merging!

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [reply tracing-opentelemetry/src/layer.rs:952]: Oh I misread, dropping the oldest events is definitely the best option here!

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs:1502]: can you also just quickly test this in one of the other tests that doesnt set `max_events_per_span`?

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs:1548]: if `max_events_per_span` is set, we could use `Vec::reserve` to set this to the exact same size, whenever we resize the vec, right? small optimization, if you are interested, not gonna block merging on it!

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [reply tracing-opentelemetry/src/layer.rs]: sounds perfect!

File: tracing-opentelemetry/src/layer.rs
Comment (guswynn) [inline tracing-opentelemetry/src/layer.rs:1553]: can you also test that this attribute doesn't show up if there are no dropped events?

**Review by guswynn — APPROVED** [review]
Looks good! Had a couple extra nits, not blocking!

File: tracing-opentelemetry/src/layer.rs
Comment (do-not-use-parker-timmerman) [reply tracing-opentelemetry/src/layer.rs:1553]: added!

File: tracing-opentelemetry/src/layer.rs
Comment (do-not-use-parker-timmerman) [reply tracing-opentelemetry/src/layer.rs:1548]: Good thought! I don't think I'm going to do this though because once we hit our max, then the Vec will already be large enough to store all of our events, so there's no need to resize, and I don't think we want to preemptively reserve enough space, because then a lot of spans would have excess capacity when they don't need it

File: tracing-opentelemetry/src/layer.rs
Comment (do-not-use-parker-timmerman) [reply tracing-opentelemetry/src/layer.rs:1502]: good idea! I added another test `exclude_max_events_per_span` which seems to follow the convention of the other `exclude_*` tests

**Pull request merged**
